src/transformers/trainer.py

```diff
@@ -3096,19 +3102,19 @@ def _determine_best_metric(self, metrics, trial):
                 ) from exc
 
             operator = np.greater if self.args.greater_is_better else np.less
-        else:
-            metric_value = metrics["eval_loss"]
-            operator = np.less
 
-        if self.state.best_metric is None or operator(metric_value, self.state.best_metric):
-            run_dir = self._get_output_dir(trial=trial)
-            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
-            output_dir = os.path.join(run_dir, checkpoint_folder)
+            if self.state.best_metric is None:
+                self.state.best_metric = float("-inf") if self.args.greater_is_better else float("inf")
+
+            if operator(metric_value, self.state.best_metric):
+                run_dir = self._get_output_dir(trial=trial)
+                checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+                output_dir = os.path.join(run_dir, checkpoint_folder)
```

Comment:
I changed the previous code of `if self.state.best_metric is None` to explicitly setting upper and lower bounds for the values for comparisons, along with separating it out from the value comparison (i.e., `operator`).

When the `metric_for_best_model` was set to a loss value, having the `self.state.best_metric` be `None` wasn't properly comparing the values.

Let me know if there'd be a better way.